server/integration/infinispan/src/main/java/org/jboss/as/clustering/infinispan/cs/factory/PromiseMap.java

```diff
@@ -0,0 +1,116 @@
+package org.jboss.as.clustering.infinispan.cs.factory;
+
+import org.jboss.as.clustering.infinispan.InfinispanLogger;
+
+import java.util.Collection;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+/**
+ * A map implementation based on promises. {@link #get(Object, long, java.util.concurrent.TimeUnit)} method will
+ * block until a corresponding {@link #put(Object, Object)} is called or a timeout occurs.
+ *
+ * @author Sebastian Laskawiec
+ * @since 8.0
+ */
+public class PromiseMap<K, V> implements Map<K, V> {
+
+   private static final TimeUnit DEFAULT_WAIT_TIME_UNIT = TimeUnit.MILLISECONDS;
+   private static final long DEFAULT_WAIT_TIMEOUT = 100;
+
+   private ConcurrentHashMap<K,CompletableFuture<V>> internalMap = new ConcurrentHashMap<>();
+
+   @Override
+   public int size() {
+      return internalMap.size();
+   }
+
+   @Override
+   public boolean isEmpty() {
+      return internalMap.isEmpty();
+   }
+
+   @Override
+   public boolean containsKey(Object key) {
+      return internalMap.containsKey(key);
+   }
+
+   @Override
+   public boolean containsValue(Object value) {
+      return true;
+   }
+
+   @Override
+   public V get(Object key) {
+      throw new UnsupportedOperationException("Use #get(Object, long, TimeUnit)");
+   }
+
+   public V get(Object key, long timeout, TimeUnit timeUnit) throws InterruptedException, ExecutionException, TimeoutException {
+      CompletableFuture<V> promise = createNewMapEntryIfNeeded((K) key);
+      return promise.get(timeout, timeUnit);
+   }
+
+   @Override
+   public V put(K key, V value) {
+      CompletableFuture<V> promise = createNewMapEntryIfNeeded(key);
+      promise.complete(value);
+      try {
+         return promise.get();
+      } catch (InterruptedException | ExecutionException ignored) {
+         //since we called promise#complete this can not block.
+         throw new AssertionError("Can not block.");
+      }
+   }
+
+   private CompletableFuture<V> createNewMapEntryIfNeeded(K key) {
+      CompletableFuture<V> newPromise = new CompletableFuture<>();
+      CompletableFuture<V> putResult = internalMap.putIfAbsent(key, newPromise);
+      if(putResult != null) {
+         return putResult;
+      }
+      return newPromise;
+   }
+
+   @Override
+   public V remove(Object key) {
+      try {
+         return internalMap.remove(key).get(DEFAULT_WAIT_TIMEOUT,DEFAULT_WAIT_TIME_UNIT);
```

Comment:
This could lead to NPE if there was no previous value to remove.  Actually waiting seems to be quite worthless as it may never ever complete as you could remove the value before the put sees the value and puts a new one.